### PR TITLE
[SPARK-48753] Upgrade arrow to v17

### PR DESCRIPTION
**Pull request by ptxmac**
Jira: https://issues.apache.org/jira/browse/SPARK-48753

### What changes were proposed in this pull request?

The current version of arrow (v12) is more than a year old. This PR simply upgrades to the most recent (v17)

### Why are the changes needed?

Arrow should be backwards compatible, so there's no reason to not use the latest stable version which comes with plenty of bug fixes

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Using `make test`

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA and link it to the PR title though https://issues.apache.org/jira/projects/SPARK/issues

File: go.mod [hunk 1]
```diff
@@ -15,46 +15,51 @@
 
 module github.com/apache/spark-connect-go/v1
 
-go 1.19
+go 1.21
```
Comment (ptxmac) [inline go.mod:18]: This assumes #20 will be accepted and merged first, otherwise we can change this back to 1.19

**Comment by grundprinzip** [comment]
Do you mind rebasing your PR to the most recent code? Thanks!

**Comment by ptxmac** [comment]
> Do you mind rebasing your PR to the most recent code? Thanks!

fun fact - in the meantime v17 was released 😂

**Review by grundprinzip — APPROVED** [review]

**Comment by grundprinzip** [comment]
Merging to master.

**Pull request closed without merging**